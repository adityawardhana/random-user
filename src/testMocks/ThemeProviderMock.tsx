import React from "react";
import theme, { GlobalStyles } from "@adityawardhana/ts-react-ui/lib/theme";
import { ThemeProvider } from "styled-components";

export const setupTheme = (children: React.ReactNode) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      {children}
    </ThemeProvider>
  );
};
