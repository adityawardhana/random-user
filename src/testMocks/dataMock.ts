export const resultAPIMock = {
  login: {
    username: "adityawrdhn",
  },
  name: {
    first: "adit",
    last: "w",
  },
  email: "adityawrdhn@gmail.com",
  gender: "male",
  registered: { date: "2003-08-26T21:56:32.575Z" },
};

export const usersMock = {
  params: { sortBy: "email", sortOrder: "ascend", page: 1, gender: "" },
  loading: false,
  users: [...Array(3)].map(() => ({ ...resultAPIMock })),
  keyword: "xxx",
  handleChangeSort: jest.fn(),
  handleChangePage: jest.fn(),
  handleChangeSearch: jest.fn(),
  setSearchFocus: jest.fn(),
  handleSelectGender: jest.fn(),
  handleResetFilter: jest.fn(),
};
