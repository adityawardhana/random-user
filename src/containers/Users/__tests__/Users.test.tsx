/* eslint-disable testing-library/no-node-access */
/* eslint-disable testing-library/no-container */
import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { setupTheme } from "../../../testMocks/ThemeProviderMock";
import User from "..";
import useFetchMock from "../../../hooks/useFetch";
import { usersMock } from "../../../testMocks/dataMock";

jest.mock("../../../hooks/useFetch", () => ({
  __esModule: true,
  default: jest.fn(),
}));
const useFetchMockValue = {
  data: { results: usersMock.users },
  loading: false,
  error: false,
  fetch: jest.fn(),
};
describe("Users unit test", () => {
  beforeEach(() => {
    (useFetchMock as any).mockReturnValue(useFetchMockValue);
  });
  test("should render users", () => {
    const { container } = render(setupTheme(<User />));
    expect(container).toMatchSnapshot();
  });

  test("should enable to type search", async () => {
    jest.useFakeTimers();
    render(setupTheme(<User />));
    const input = screen.getAllByRole("textbox")[0];
    fireEvent.focus(input);
    fireEvent.input(input, {
      target: { value: "type" },
    });

    jest.runOnlyPendingTimers();
    await waitFor(
      () => {
        expect(useFetchMockValue.fetch).toHaveBeenCalled();
      },
      { timeout: 1000 }
    );
  });

  test("should open dropdown and select gender", async () => {
    const { container } = render(setupTheme(<User />));

    fireEvent.click(screen?.getAllByRole("button")[0]);
    fireEvent.click(container.querySelectorAll("li")[1]);
    expect(container.querySelectorAll("button")[0].textContent).toBe("Male");
  });

  test("should enable to reset filter", async () => {
    render(setupTheme(<User />));
    fireEvent.click(screen.getByText(/Reset Filter/i));
  });

  test("should enable to sort table", async () => {
    render(setupTheme(<User />));
    fireEvent.click(screen.getByText(/Username/i));
  });

  test("should enable to change page", async () => {
    render(setupTheme(<User />));
    const allButtons = screen?.getAllByRole("button");
    fireEvent.click(allButtons[allButtons.length - 1]);
  });
});
