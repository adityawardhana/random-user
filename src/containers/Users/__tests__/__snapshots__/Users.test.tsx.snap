// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users unit test should render users 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-bottom: 40px;
}

.c1 {
  width: 200px;
  margin-right: 16px;
}

.c6 {
  position: relative;
}

.c9 {
  width: 200px;
}

.c13 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c15 {
  margin-top: 20px;
  margin-bottom: 20px;
}

.c16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c8 {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c14 {
  font-size: 12px;
  font-weight: 400;
  line-height: 16px;
  color: #152E3B;
  white-space: normal;
  font-weight: bold;
}

.c19 {
  font-size: 16px;
  font-weight: 400;
  line-height: 20px;
  margin-right: 8px;
  margin-left: 8px;
  color: #243E47;
  white-space: normal;
}

.c5 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 16px;
  height: 16px;
  pointer-events: none;
}

.c5[type="button"] .c4 {
  cursor: pointer;
}

.c5 svg {
  width: 16px;
  height: 16px;
  fill: #152E3B;
}

.c18 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 16px;
  height: 16px;
  pointer-events: none;
}

.c18[type="button"] .c4 {
  cursor: pointer;
}

.c18 svg {
  width: 16px;
  height: 16px;
  fill: #A6A8AF;
}

.c20 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 16px;
  height: 16px;
  pointer-events: none;
}

.c20[type="button"] .c4 {
  cursor: pointer;
}

.c20 svg {
  width: 16px;
  height: 16px;
  fill: #243E47;
}

.c7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  overflow: hidden;
  border: 1px solid transparent;
  border-radius: 8px;
  border-color: transparent;
  background-color: #F4F5F8;
  color: #152E3B;
  padding: 12px 16px;
  height: 40px;
  width: auto;
  font-family: inherit;
  font-size: 14px;
  font-weight: 500;
  text-align: center;
  vertical-align: middle;
  white-space: nowrap;
  cursor: pointer;
  box-sizing: border-box;
  -webkit-transition: -webkit-transform 0.1s ease 0s,background 0.3s ease !important;
  -webkit-transition: transform 0.1s ease 0s,background 0.3s ease !important;
  transition: transform 0.1s ease 0s,background 0.3s ease !important;
  width: 100%;
}

.c7::before {
  content: "";
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c7:hover::before {
  background: rgba(255,255,255,0.1);
}

.c7:focus {
  outline: none;
}

.c7:focus-visible::before {
  background: rgba(50,50,50,0.05);
  -webkit-transition: background 0.1s ease-in-out;
  transition: background 0.1s ease-in-out;
}

.c7:active {
  outline: none;
  -webkit-transform: scale(0.96) !important;
  -ms-transform: scale(0.96) !important;
  transform: scale(0.96) !important;
}

.c7:active::before {
  background: rgba(50,50,50,0.05);
  -webkit-transition: background 0.3s ease;
  transition: background 0.3s ease;
}

.c7:disabled {
  color: #CDD0D7;
  cursor: default;
}

.c7:link,
.c7:visited,
.c7:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7[type="button"],
.c7[type="reset"],
.c7[type="submit"] {
  -webkit-appearance: none;
}

.c10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  overflow: hidden;
  border: 1px solid transparent;
  border-radius: 8px;
  border-color: #152E3B;
  background-color: transparent;
  color: #152E3B;
  padding: 12px 16px;
  height: 40px;
  width: auto;
  font-family: inherit;
  font-size: 14px;
  font-weight: 500;
  text-align: center;
  vertical-align: middle;
  white-space: nowrap;
  cursor: pointer;
  box-sizing: border-box;
  -webkit-transition: -webkit-transform 0.1s ease 0s,background 0.3s ease !important;
  -webkit-transition: transform 0.1s ease 0s,background 0.3s ease !important;
  transition: transform 0.1s ease 0s,background 0.3s ease !important;
}

.c10::before {
  content: "";
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c10:hover::before {
  background: rgba(255,255,255,0.1);
}

.c10:focus {
  outline: none;
}

.c10:focus-visible::before {
  background: rgba(50,50,50,0.05);
  -webkit-transition: background 0.1s ease-in-out;
  transition: background 0.1s ease-in-out;
}

.c10:active {
  outline: none;
  -webkit-transform: scale(0.96) !important;
  -ms-transform: scale(0.96) !important;
  transform: scale(0.96) !important;
}

.c10:active::before {
  background: rgba(50,50,50,0.05);
  -webkit-transition: background 0.3s ease;
  transition: background 0.3s ease;
}

.c10:disabled {
  border-color: #CDD0D7;
  color: #CDD0D7;
  cursor: default;
}

.c10:link,
.c10:visited,
.c10:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c10[type="button"],
.c10[type="reset"],
.c10[type="submit"] {
  -webkit-appearance: none;
}

.c17 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  overflow: hidden;
  border: 1px solid transparent;
  border-radius: 50%;
  border-color: transparent;
  background-color: transparent;
  color: #152E3B;
  padding: 8px;
  height: 28px;
  width: 28px;
  font-family: inherit;
  font-size: 10px;
  font-weight: 500;
  text-align: center;
  vertical-align: middle;
  white-space: nowrap;
  cursor: pointer;
  box-sizing: border-box;
  -webkit-transition: -webkit-transform 0.1s ease 0s,background 0.3s ease !important;
  -webkit-transition: transform 0.1s ease 0s,background 0.3s ease !important;
  transition: transform 0.1s ease 0s,background 0.3s ease !important;
}

.c17::before {
  content: "";
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.c17:hover::before {
  background: rgba(255,255,255,0.1);
}

.c17:focus {
  outline: none;
}

.c17:focus-visible::before {
  background: rgba(50,50,50,0.05);
  -webkit-transition: background 0.1s ease-in-out;
  transition: background 0.1s ease-in-out;
}

.c17:active {
  outline: none;
  -webkit-transform: scale(0.96) !important;
  -ms-transform: scale(0.96) !important;
  transform: scale(0.96) !important;
}

.c17:active::before {
  background: rgba(50,50,50,0.05);
  -webkit-transition: background 0.3s ease;
  transition: background 0.3s ease;
}

.c17:disabled {
  color: #CDD0D7;
  cursor: default;
}

.c17:link,
.c17:visited,
.c17:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c17[type="button"],
.c17[type="reset"],
.c17[type="submit"] {
  -webkit-appearance: none;
}

.c2 {
  width: 100%;
  box-sizing: border-box;
  position: relative;
  display: block;
  -webkit-transition: all 0.5s ease;
  transition: all 0.5s ease;
}

.c2 .icon-search {
  top: 12px;
  right: 12px;
  position: absolute;
}

.c3 {
  box-shadow: none;
  margin-left: 0;
  height: 40px !important;
  border-color: transparent;
  background-color: #F4F5F8;
  color: #243E47;
  padding: 10px 34px 10px 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  box-sizing: border-box;
  font-size: 14px;
  line-height: 20px;
  border-radius: 8px;
  -webkit-transition: all 0.5s ease;
  transition: all 0.5s ease;
}

.c11 {
  display: block;
  width: 100%;
  overflow-x: auto;
  border-radius: 4px;
  -moz-border-radius: 4px;
  -webkit-border-radius: 4px;
}

.c11.shadow {
  box-shadow: 0 1px 4px 0 rgba(112,114,125,0.4);
}

.c11 table {
  width: 100%;
}

.c11::-webkit-scrollbar {
  height: 8px;
}

.c11::-webkit-scrollbar-thumb {
  background-color: #dddddd;
  border-radius: 10px;
}

.c11::-webkit-scrollbar-track {
  background-color: #e6e6e659;
}

.c11 {
  box-shadow: 0 1px 4px 0 rgba(112,114,125,0.4);
}

.c12 {
  background-color: #ffffff;
  border-radius: 4px;
  -moz-border-radius: 4px;
  -webkit-border-radius: 4px;
  border-collapse: separate;
}

.c12 thead th {
  background-color: #ffffff;
  color: #70727d;
}

.c12 tbody tr {
  background-color: #f8fbfc;
}

.c12 tbody tr:nth-child(2n) {
  background-color: #ffffff;
}

.c12 th:first-of-type {
  border-radius: 4px 0 0 0;
  -moz-border-radius: 4px 0 0 0;
  -webkit-border-radius: 4px 0 0 0;
}

.c12 th:last-of-type {
  border-radius: 0 4px 0 0;
  -moz-border-radius: 0 4px 0 0;
  -webkit-border-radius: 0 4px 0 0;
}

.c12 tr:last-of-type td:first-of-type {
  border-radius: 0 0 0 4px;
  -moz-border-radius: 0 0 0 4px;
  -webkit-border-radius: 0 0 0 4px;
}

.c12 tr:last-of-type td:last-of-type {
  border-radius: 0 0 4px 0;
  -moz-border-radius: 0 0 4px 0;
  -webkit-border-radius: 0 0 4px 0;
}

.c12 td:first-of-type,
.c12 th:first-of-type {
  border-left: medium none;
}

.c12 th,
.c12 td {
  padding: 14px 24px;
  font-size: 12px;
  vertical-align: middle;
}

.c12 th[rowspan],
.c12 td[rowspan] {
  vertical-align: middle;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      width="200"
    >
      <div
        class="c2"
      >
        <input
          class="c3"
          placeholder="Type your search here..."
          value=""
        />
        <i
          class="c4 c5 icon-search"
          color="secondary"
        >
          <svg
            aria-hidden="true"
            class="octicon octicon-search"
            fill="currentColor"
            height="16"
            role="img"
            style="display: inline-block; user-select: none; vertical-align: text-bottom;"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M11.5 7a4.499 4.499 0 11-8.998 0A4.499 4.499 0 0111.5 7zm-.82 4.74a6 6 0 111.06-1.06l3.04 3.04a.75.75 0 11-1.06 1.06l-3.04-3.04z"
              fill-rule="evenodd"
            />
          </svg>
        </i>
      </div>
    </div>
    <div
      class="c1"
      width="200"
    >
      <div
        class="c6"
      >
        <button
          class="c7"
          color="secondary"
          shape="rectangle"
          style="justify-content: space-between;"
        >
          <div
            class="c8"
          >
            <s
              style="position: absolute; border-radius: 50%; opacity: 0; width: 35px; height: 35px; transform: translate(-50%, -50%); pointer-events: none; background-color: rgba(255, 255, 255, 0.4);"
            />
          </div>
          All
          <i
            class="c4 c5"
            color="secondary"
          >
            <svg
              aria-hidden="true"
              class="octicon octicon-chevron-down"
              fill="currentColor"
              height="16"
              role="img"
              style="display: inline-block; user-select: none; vertical-align: text-bottom;"
              viewBox="0 0 16 16"
              width="16"
            >
              <path
                d="M12.78 6.22a.75.75 0 010 1.06l-4.25 4.25a.75.75 0 01-1.06 0L3.22 7.28a.75.75 0 011.06-1.06L8 9.94l3.72-3.72a.75.75 0 011.06 0z"
                fill-rule="evenodd"
              />
            </svg>
          </i>
        </button>
      </div>
    </div>
    <div
      class="c9"
      width="200"
    >
      <button
        class="c10"
        color="secondary"
        shape="rectangle"
      >
        <div
          class="c8"
        >
          <s
            style="position: absolute; border-radius: 50%; opacity: 0; width: 35px; height: 35px; transform: translate(-50%, -50%); pointer-events: none; background-color: rgba(255, 255, 255, 0.4);"
          />
        </div>
        Reset Filter
      </button>
    </div>
  </div>
  <div
    class="c11"
  >
    <table
      class="c12"
    >
      <thead>
        <tr>
          <th>
            <span
              class="c13"
              cursor="pointer"
            >
              <span
                class="c14"
                color="black900"
                type="B12"
              >
                Username
              </span>
            </span>
          </th>
          <th>
            <span
              class="c13"
              cursor="pointer"
            >
              <span
                class="c14"
                color="black900"
                type="B12"
              >
                Name
              </span>
            </span>
          </th>
          <th>
            <span
              class="c13"
              cursor="pointer"
            >
              <span
                class="c14"
                color="black900"
                type="B12"
              >
                Email
              </span>
            </span>
          </th>
          <th>
            <span
              class="c13"
              cursor="pointer"
            >
              <span
                class="c14"
                color="black900"
                type="B12"
              >
                Gender
              </span>
            </span>
          </th>
          <th>
            <span
              class="c13"
              cursor="pointer"
            >
              <span
                class="c14"
                color="black900"
                type="B12"
              >
                Registered Date
              </span>
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            adityawrdhn
          </td>
          <td>
            adit
             
            w
          </td>
          <td>
            adityawrdhn@gmail.com
          </td>
          <td>
            male
          </td>
          <td>
            27-08-2003 04:56
          </td>
        </tr>
        <tr>
          <td>
            adityawrdhn
          </td>
          <td>
            adit
             
            w
          </td>
          <td>
            adityawrdhn@gmail.com
          </td>
          <td>
            male
          </td>
          <td>
            27-08-2003 04:56
          </td>
        </tr>
        <tr>
          <td>
            adityawrdhn
          </td>
          <td>
            adit
             
            w
          </td>
          <td>
            adityawrdhn@gmail.com
          </td>
          <td>
            male
          </td>
          <td>
            27-08-2003 04:56
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="c15"
  >
    <div
      class="c16"
    >
      <button
        class="c17"
        color="secondary"
        shape="circle"
      >
        <div
          class="c8"
        >
          <s
            style="position: absolute; border-radius: 50%; opacity: 0; width: 35px; height: 35px; transform: translate(-50%, -50%); pointer-events: none; background-color: rgba(255, 255, 255, 0.4);"
          />
        </div>
        <i
          class="c4 c18"
          color="black400"
        >
          <svg
            aria-hidden="true"
            class="octicon octicon-chevron-left"
            fill="currentColor"
            height="16"
            role="img"
            style="display: inline-block; user-select: none; vertical-align: text-bottom;"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M9.78 12.78a.75.75 0 01-1.06 0L4.47 8.53a.75.75 0 010-1.06l4.25-4.25a.75.75 0 011.06 1.06L6.06 8l3.72 3.72a.75.75 0 010 1.06z"
              fill-rule="evenodd"
            />
          </svg>
        </i>
      </button>
      <p
        class="c19"
        color="black800"
        type="B16"
      >
        1
      </p>
      <button
        class="c17"
        color="secondary"
        shape="circle"
      >
        <div
          class="c8"
        >
          <s
            style="position: absolute; border-radius: 50%; opacity: 0; width: 35px; height: 35px; transform: translate(-50%, -50%); pointer-events: none; background-color: rgba(255, 255, 255, 0.4);"
          />
        </div>
        <i
          class="c4 c20"
          color="black800"
        >
          <svg
            aria-hidden="true"
            class="octicon octicon-chevron-right"
            fill="currentColor"
            height="16"
            role="img"
            style="display: inline-block; user-select: none; vertical-align: text-bottom;"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M6.22 3.22a.75.75 0 011.06 0l4.25 4.25a.75.75 0 010 1.06l-4.25 4.25a.75.75 0 01-1.06-1.06L9.94 8 6.22 4.28a.75.75 0 010-1.06z"
              fill-rule="evenodd"
            />
          </svg>
        </i>
      </button>
    </div>
  </div>
</div>
`;
